src/proxy-ui-api/src/main/java/org/niis/xroad/restapi/openapi/ClientsApiController.java

```diff
@@ -336,9 +341,25 @@ private ClientType getClientType(String encodedId) {
                 // deviation data (errorcode + warnings) copied
                 throw new ConflictException(e);
             }
-
         } else if (serviceDescription.getType() == ServiceType.OPENAPI3) {
-            return new ResponseEntity<>(HttpStatus.NOT_IMPLEMENTED);
+            try {
+                addedServiceDescriptionType = serviceDescriptionService.addOpenapi3ServiceDescription(clientId, url,
+                        restServiceCode, ignoreWarnings);
+            } catch (OpenApiParser.ParsingException | UnhandledWarningsException e) {
+                throw new BadRequestException(e);
+            } catch (ClientNotFoundException e) {
+                throw new ResourceNotFoundException(e);
+            } catch (ServiceDescriptionService.UrlAlreadyExistsException
+                    | ServiceDescriptionService.ServiceCodeAlreadyExistsException e) {
+                throw new ConflictException(e);
+            }
+        } else if (serviceDescription.getType() == ServiceType.REST) {
+            try {
+                addedServiceDescriptionType = serviceDescriptionService.addRestEndpointServiceDescription(clientId,
+                        url, serviceDescription.getRestServiceCode());
```

Comment:
`serviceDescription.getRestServiceCode()` might return null here which causes problems later. Needs a null check and proper exception (BadRequestException). Could also use the above introduced `restServiceCode` variable.